minijinja/src/value/mod.rs

```diff
@@ -593,13 +606,41 @@ impl Value {
     /// ```
     pub fn as_slice(&self) -> Result<&[Value], Error> {
         match self.0 {
-            ValueRepr::Undefined | ValueRepr::None => Ok(&[][..]),
-            ValueRepr::Seq(ref v) => Ok(&v[..]),
-            _ => Err(Error::new(
-                ErrorKind::InvalidOperation,
-                format!("value of type {} is not a sequence", self.kind()),
-            )),
+            ValueRepr::Undefined | ValueRepr::None => return Ok(&[][..]),
+            ValueRepr::Seq(ref v) => return Ok(&v[..]),
+            ValueRepr::Dynamic(ref dy) if matches!(dy.behavior(), ObjectBehavior::Seq(_)) => {
+                return Err(Error::new(
+                    ErrorKind::InvalidOperation,
+                    "dynamic sequence value cannot be borrowed as slice",
+                ));
+            }
+            _ => {}
+        }
+        Err(Error::new(
+            ErrorKind::InvalidOperation,
+            format!("value of type {} is not a sequence", self.kind()),
+        ))
+    }
+
+    /// If the value is a sequence it's returned as slice.
+    ///
+    /// ```
+    /// # use minijinja::value::Value;
+    /// let seq = Value::from(vec![1u32, 2, 3, 4]);
+    /// let slice = seq.as_slice().unwrap();
+    /// assert_eq!(slice.len(), 4);
+    /// ```
+    pub(crate) fn as_cow_slice(&self) -> Result<Cow<'_, [Value]>, Error> {
+        if let ValueRepr::Dynamic(ref dy) = self.0 {
+            if let ObjectBehavior::Seq(seq) = dy.behavior() {
+                let mut items = vec![];
+                for idx in 0..seq.len() {
+                    items.push(seq.get(idx).unwrap_or(Value::UNDEFINED));
+                }
+                return Ok(Cow::Owned(items));
+            }
         }
+        self.as_slice().map(Cow::Borrowed)
```

Comment:
It's odd to me that this one works with the dynamic `Seq` while `as_slice()` doesn't since the comment in `as_slice()` says that dynamic objects aren't and can't be slices. This method is also expensive and effectively undoes the work of putting something into an `AsSeq` in the first place. I really think returning a `&dyn AsSeq` is much more advantageous here.

Reply:
`as_cow_slice` is internal and only exists in places where the entire slice is needed. Eg it's used for the `reverse` filter or argument unpacking so changing to `AsSeq` is not particularly useful.